examples/server/server.cpp

```diff
@@ -2173,6 +2177,14 @@ struct server_context {
                         }
                     }
 
+                    // check that we are in the right batch_type, if not defer the slot
+                    bool slot_type = slot.embedding ? 1 : 0;
+                    if (batch_type == -1) {
+                        batch_type = slot_type;
+                    } else if (batch_type != slot_type) {
+                        continue;
+                    }
```

Comment:
Should this be done earlier? There's some things like `slot.n_past` which should probably not be changed when the batch_type is not the right one.

I feel like this check should be done at least before tokens are put in the `batch`, but the multitude of loops over slots does make it hard to find one right spot.

Reply:
Indeed. But we also want to make sure we don't assign the batch to a particular type but then bail out later on (say, because the prompt is bigger than the batch size).

My current read is that everything before where the check is currently is tokenization, and it's fine to do that, bail, and then pick up the slot on the next go around. That's why it's okay to have the conditional `continue` just above it (2173-2178) too.

Reply:
> everything before where the check is currently is tokenization and it's fine to do that, bail, and then pick up the slot on the next go around.

Right, and `slot.n_past` is recalculated anyway, so you're right.

But I think if `batch.n_tokens` is non-zero, the initial `batch_type` should be `non-embedding`, or else this could potentially attempt continuous batching with embeddings *and* previous text completions.

Reply:
Ah I see, just pushed the fix!